### DISCUSSION TITLE
Possible to create new files?

### DISCUSSION THREAD

**Discussion by FinnNk** [sn4k3/UVtools#324] [Scripts]
It seems that to create a new file I first need a dummy or host file which I can then manipulate. Is this correct or is it possible to create new files (CTB v3 in what I'm attempting) from a set of layer mats?

**Comment by tslater2006** [comment]


**Reply by sn4k3** [reply]
They are not a problem, UVtools will auto populate those by default based on observations. Renember there are people using PrusaSlicer and convert to CTB and printing with success. Most of unknowns are info only and not required anyway, but in most case UVtools will handle it by default

**Comment by sn4k3** [comment]
It is possible to create new files.

```C#
var slicerFile = new ChituboxFile();
// Populate all settings here: resolution, display size, layer height, bottom count, etc
slicerFile.Resolution = size; //eg
slicerFile.Display = size; //eg
slicerFile.LayerHeight = 0.05f; //eg
// ...
// ...
slicerFile.LayerManager.AllocateAndSetFromMat(mat[]);  // Set from a mat list
// OR:
slicerFile.LayerManager.Init(count or layers[]); // Init from a count or a layer array, decodes use this method with number and init each layer
// Then:
slicerFile.SaveAs("my_new_file.ctb");
```

Dont forget that you will need to populate all other settings, such as resolution, display, etc before hand that's why load a dummy model is easier.

Note: The **Init** method will not notify file from layer changes and should only be used when reading back a file that don't require recaulculations or resets, use **AllocateAndSetFromMat** is safer and will auto compute some fields. You can also do `slicerFile.LayerManager.Layers = myLayerSet;` if you wish to pass a **Layer[]** instead of **Mat[]**, but using a Mat array is simpler (If you have the RAM to hold such list) and UVtools will populate all layer data for you based on global settings.
If RAM is a problem you can also Surpress notifications, **Init** the layers with a count, set the Mat for each layer, Unsurpress notifications, force recalculation and notify of change, eg:

```C#
SlicerFile.SuppressRebuildPropertiesWork(() =>
            {
                SlicerFile.LayerHeight = (float)LayerHeight;
                SlicerFile.BottomExposureTime = (float)BottomExposure;
                SlicerFile.ExposureTime = (float)NormalExposure;
                SlicerFile.BottomLayerCount = BottomLayers;

                SlicerFile.LayerManager.Init(count);

                // Parallel loop each created/loaded Mat and and set to layer index
                Parallel.For(0, count, ....) ....
            }, true);
```

That method will mute all notifications, and when exit will recalculate and sanitize everything

**Comment by FinnNk** [comment]
Perfect - thanks!
